### DISCUSSION TITLE
pci device can't hotplug  for aarch64

### DISCUSSION THREAD

**Discussion by grass-lu** [cloud-hypervisor/cloud-hypervisor#5910] [Q&A] [answered]
**Describe the bug**

pci device can't hotplug  for aarch64, and can't hotplug memory

**To Reproduce**
Steps to reproduce the behaviour:

1、start vm
./cloud-hypervisor --kernel ./Image --console off --serial tty --disk path=/home/luzhipeng/openEuler-23.09-aarch64.raw --cmdline "console=ttyAMA0 root=/dev/vda2 rw" --cpus boot=4,max=16 --memory size=1024M,hotplug_size=8192M --net tap=,mac=,ip=,mask= --api-socket /tmp/cloud.sock --log-file /tmp/cloud-hypervisor.log -vv --seccomp false

2、 look up  disk list of vm

[root@cloud-hypervisor ~]# lsblk
NAME   MAJ:MIN RM  SIZE RO TYPE MOUNTPOINTS
nullb0 252:0    0  250G  0 disk
zram0  253:0    0    0B  0 disk
vda    254:0    0   40G  0 disk
├─vda1 254:1    0    2G  0 part /boot
└─vda2 254:2    0   38G  0 part /

3、 hotplug disk
[root@wlcjs172-master-03 cloud-hypervisor]# ./ch-remote --api-socket=/tmp/cloud.sock   add-disk path=cloud.img
{"id":"_disk2","bdf":"0000:00:04.0"}

4、 look up  disk list of vm
[root@cloud-hypervisor ~]# lsblk
NAME   MAJ:MIN RM  SIZE RO TYPE MOUNTPOINTS
nullb0 252:0    0  250G  0 disk
zram0  253:0    0    0B  0 disk
vda    254:0    0   40G  0 disk
├─vda1 254:1    0    2G  0 part /boot
└─vda2 254:2    0   38G  0 part /

**Version**

 ./cloud-hypervisor --version
cloud-hypervisor v35.0-84-g4cbfccc1

Output of `cloud-hypervisor --version`:

Did you build from source, if so build command line (e.g. features):

build command line:

cargo build --release

**VM configuration**

./cloud-hypervisor --kernel ./Image --console off --serial tty --disk path=/home/luzhipeng/openEuler-23.09-aarch64.raw --cmdline "console=ttyAMA0 root=/dev/vda2 rw" --cpus boot=4,max=16 --memory size=1024M,hotplug_size=8192M --net tap=,mac=,ip=,mask= --api-socket /tmp/cloud.sock --log-file /tmp/cloud-hypervisor.log -vv --seccomp false

What command line did you run (or JSON config data):

Guest OS version details:
[root@cloud-hypervisor ~]# uname -ar
Linux cloud-hypervisor 6.2.0+ #1 SMP Mon Oct 30 19:11:01 CST 2023 aarch64 aarch64 aarch64 GNU/Linux

**Guest os config**
cat /boot/config-6.4.0-10.1.0.20.oe2309.aarch64 | gre
CONFIG_HOTPLUG_CPU=y
CONFIG_ACPI_HOTPLUG_CPU=y
CONFIG_ACPI_HOTPLUG_MEMORY=y
CONFIG_ARCH_ENABLE_MEMORY_HOTPLUG=y
CONFIG_MEMORY_HOTPLUG=y
CONFIG_MEMORY_HOTPLUG_DEFAULT_ONLINE=y
CONFIG_HOTPLUG_PCI_PCIE=y
CONFIG_HOTPLUG_PCI=y
CONFIG_HOTPLUG_PCI_ACPI=y
CONFIG_HOTPLUG_PCI_ACPI_IBM=m
CONFIG_HOTPLUG_PCI_SHPC=y
CONFIG_XEN_BALLOON_MEMORY_HOTPLUG=y
CONFIG_SURFACE_HOTPLUG=m

Host OS version details:

[root@wlcjs172-master-03 cloud-hypervisor]# uname -ar
Linux wlcjs172-master-03 5.15.67-6.cl9.aarch64 #1 SMP Wed Mar 8 06:37:12 UTC 2023 aarch64 aarch64 aarch64 GNU/Linux

**Logs**

Output of `cloud-hypervisor -v` from either standard error or via `--log-file`:

cloud-hypervisor: 81.011850s: <vmm> INFO:vmm/src/lib.rs:1952 -- API request event: VmAddDisk(DiskConfig { path: Some("cloud.img"), readonly: false, direct: false, iommu: false, num_queues: 1, queue_size: 128, vhost_user: false, vhost_socket: None, rate_limiter_config: None, id: None, disable_io_uring: false, disable_aio: false, pci_segment: 0, serial: None }, Sender { .. })

cloud-hypervisor: 81.012101s: <vmm> INFO:vmm/src/device_manager.rs:2212 -- Creating virtio-block device: DiskConfig { path: Some("cloud.img"), readonly: false, direct: false, iommu: false, num_queues: 1, queue_size: 128, vhost_user: false, vhost_socket: None, rate_limiter_config: None, id: Some("_disk2"), disable_io_uring: false, disable_aio: false, pci_segment: 0, serial: None }

cloud-hypervisor: 81.012381s: <vmm> INFO:vmm/src/device_manager.rs:2302 -- Using asynchronous RAW disk file (io_uring)

cloud-hypervisor: 81.012425s: <vmm> INFO:virtio-devices/src/block.rs:584 -- Disk topology: DiskTopology { logical_block_size: 512, physical_block_size: 512, minimum_io_size: 512, optimal_io_size: 0 }

Linux kernel output:

**Comment by grass-lu** [comment]
@fdr  @bryteise @iggy @weltling

**Comment by MrXinWang** [comment, marked as answer]
See the notes in https://github.com/cloud-hypervisor/cloud-hypervisor/blob/main/docs/hotplug.md#pci-device-hot-plug, the Arm PCI device hotplug is only available if you use ACPI, however from your command line attached above, it seems that this is not the case.
